oxenstored/connections.ml

```diff
@@ -57,23 +66,39 @@ let add_domain cons dom =
     )
   in
   let con = Connection.create xbcon (Some dom) in
-  Hashtbl.add cons.domains (Domain.get_id dom) con ;
-  Hashtbl.add cons.ports (Domain.get_local_port dom) con
-
-let select ?(only_if = fun _ -> true) cons =
-  Hashtbl.fold
-    (fun _ con (ins, outs) ->
-      if only_if con then
-        let fd = Connection.get_fd con in
-        let in_fds = if Connection.can_input con then fd :: ins else ins in
-        let out_fds = if Connection.has_output con then fd :: outs else outs in
-        (in_fds, out_fds)
-      else
-        (ins, outs)
+  Hashtbl.replace cons.domains (Domain.get_id dom) con ;
+  Hashtbl.replace cons.ports (Domain.get_local_port dom) con
+
+let refresh_poll_status ?(only_if = fun _ -> true) cons spec_fds =
+  (* special fds are always read=true, but get overwritten by select_stubs, so we
+     need to reset the event we are polling for *)
+  let _ =
+    List.fold_left
+      (fun index fd ->
+        cons.poll_status.(index) <- (fd, spec_poll_status ()) ;
+        index + 1
+      )
+      0 spec_fds
+  in
+  Hashtbl.iter
+    (fun _ (con, index) ->
+      let only = only_if con in
+      let fd = Connection.get_fd con in
+      let open Poll in
+      let event =
+        {
+          read= only && Connection.can_input con
```

Comment:
As a further optimization we could update the entry in the poll table only when 'can_input' or 'has_output' changes its value (i.e. by the functions that change that value), and then we don't need to loop over all the anon connections all the time.
But that might be more difficult to get right.

Reply:
That was how I first went about it, but I think it's quite fragile and I felt unsure I had a full grasp of all potential cases where this could change.